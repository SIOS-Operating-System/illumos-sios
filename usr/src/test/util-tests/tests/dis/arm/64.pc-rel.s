/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2022 OmniOS Community Edition (OmniOSce) Association.
 */

/*
 * The assembled file must be linked in order to resolve the relocations for
 * a proper test. The following token enables that.
 * %DIS_TEST_LINK
 */

/*
 * C3.4.6 PC-rel. addressing
 */

.text

// Add some padding so that the instructions are not in page 0.
.pad:
	.space 0xc000

.globl libdis_test
.type libdis_test, @function
libdis_test:
	adr	x0, .lword
	adr	x1, .ldata
	adr	x2, .ldata + 0xffe

	adrp	x3, .lword
	add	x3, x3, :lo12:.lword

	adrp	x4, .ldata
	add	x4, x4, :lo12:.ldata

	adrp	x5, .ldata + 0xffe
	add	x5, x5, :lo12:.ldata + 0xffe

	adrp	x6, .sym
	add	x6, x6, :lo12:.sym

	adrp	x7, :got:.sym
	ldr	x7, [x7, :got_lo12:.sym]

.size libdis_test, [.-libdis_test]

// padding so that the data is not in the same page.
.space 0x17f6

.lword:	.word	0xdeadbeef
.ldata:	.xword	0x1122334455667788
	.space	0x1ffb

.global .sym
.sym:

