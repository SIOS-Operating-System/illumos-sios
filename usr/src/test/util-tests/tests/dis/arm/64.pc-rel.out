    libdis_test:      40 c1 00 50  adr x0, 0x40dd32 
    libdis_test+0x4:  41 c1 00 50  adr x1, 0x40dd36 
    libdis_test+0x8:  22 41 01 10  adr x2, 0x40ed34 
    libdis_test+0xc:  03 00 00 b0  adrp x3, 0x40d000
    libdis_test+0x10: 63 c8 34 91  add x3, x3, #0xd32
    libdis_test+0x14: 04 00 00 b0  adrp x4, 0x40d000
    libdis_test+0x18: 84 d8 34 91  add x4, x4, #0xd36
    libdis_test+0x1c: 05 00 00 d0  adrp x5, 0x40e000
    libdis_test+0x20: a5 d0 34 91  add x5, x5, #0xd34
    libdis_test+0x24: 06 00 00 f0  adrp x6, 0x40f000
    libdis_test+0x28: c6 e4 34 91  add x6, x6, #0xd39	<_etext>
    libdis_test+0x2c: 87 00 00 f0  adrp x7, 0x41f000
    libdis_test+0x30: e7 ac 46 f9  ldr x7, [x7, #3416]
